### PR TITLE
[incubator-kie-issues#1474] Fix NPE when loading multiple Signavio MID

### LINKED ISSUE #1474: NPE thrown when loading multiple DMN XML files including a Signavio's multi instance decision
## Current behaviour
Currently when trying to load multiple DMN in a certain order, where an XML file that contain a MID (Multi Instance Decision) loads first, then load another XML file. We get the following exception:

```
[main] ERROR o.k.d.core.compiler.DMNCompilerImpl.compile:200 - Error compiling model from source.
java.lang.NullPointerException: Cannot invoke "org.kie.dmn.core.ast.DMNBaseNode.getDependencies()" because "node" is null
	at org.kie.dmn.core.internal.utils.DRGAnalysisUtils.internalDependencies(DRGAnalysisUtils.java:60)
	at org.kie.dmn.core.internal.utils.DRGAnalysisUtils.dependencies(DRGAnalysisUtils.java:54)
	at org.kie.dmn.signavio.MultiInstanceDecisionLogic$MIDDependenciesProcessor.findAllDependencies(MultiInstanceDecisionLogic.java:176)
	at org.kie.dmn.signavio.MultiInstanceDecisionLogic$MultiInstanceDecisionNodeCompiler.addRequiredDecisions(MultiInstanceDecisionLogic.java:147)
	at org.kie.dmn.signavio.MultiInstanceDecisionLogic$MultiInstanceDecisionNodeCompiler.lambda$compileEvaluator$1(MultiInstanceDecisionLogic.java:139)
	at org.kie.dmn.core.compiler.DMNCompilerImpl.processDrgElements(DMNCompilerImpl.java:482)
	at org.kie.dmn.core.compiler.DMNCompilerImpl.compile(DMNCompilerImpl.java:263)
	at org.kie.dmn.core.compiler.DMNCompilerImpl.compile(DMNCompilerImpl.java:198)
	at org.kie.dmn.core.assembler.DMNAssemblerService.compileResourceToModel(DMNAssemblerService.java:177)
	at org.kie.dmn.core.assembler.DMNAssemblerService.internalAddResource(DMNAssemblerService.java:152)
	at org.kie.dmn.core.assembler.DMNAssemblerService.addResourcesAfterRules(DMNAssemblerService.java:127)
	at org.kie.internal.services.KieAssemblersImpl.addResourcesAfterRules(KieAssemblersImpl.java:86)
	at org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.buildAssemblerResourcesAfterRules(CompositeKnowledgeBuilderImpl.java:204)
	at org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.build(CompositeKnowledgeBuilderImpl.java:124)
	at org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.build(CompositeKnowledgeBuilderImpl.java:109)
	at org.drools.compiler.kie.builder.impl.AbstractKieProject.buildKnowledgePackages(AbstractKieProject.java:280)
	at org.drools.compiler.kie.builder.impl.AbstractKieProject.buildKnowledgePackages(AbstractKieProject.java:220)
	at org.drools.compiler.kie.builder.impl.AbstractKieProject.verify(AbstractKieProject.java:84)
	at org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildKieProject(KieBuilderImpl.java:285)
	at org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildAll(KieBuilderImpl.java:251)
	at org.drools.compiler.kie.builder.impl.KieBuilderImpl.buildAll(KieBuilderImpl.java:198)
```

### What, after some debug, I see is happening:
After the introduction of the `afterDRGcallbacks` in the `DMNCompilerImpl`. The `MultiInstanceDecisionLogic` decided to use the callbacks. Which adds a callback to the `DMNCompilerImpl` which in turn executes this callback regardless of which model it is evaluating currently. This lead to the evaluation of some callbacks, which were originally registered/added for the MID on some other models. This leads to a NPE when trying to execute the call back on a different model. 

## Expected behaviour
Load multiple models from XML that contain MIDs and non MIDs in a any order, then we are able to build a `KieBuilder` for them all.

## How to reproduce?
Load multiple files with Mids into a single runtime:
```java
public void createRuntime() {  
    final KieServices ks = KieServices.Factory.get();  
    final KieFileSystem kfs = ks.newKieFileSystem();  
  
    KieModuleModel kmm = ks.newKieModuleModel();  
    kmm.setConfigurationProperty("org.kie.dmn.profiles.signavio", "org.kie.dmn.signavio.KieDMNSignavioProfile");  
    kfs.writeKModuleXML(kmm.toXML());

    // Where here both files contain Signavio MIDs 
    kfs.write(ks.getResources().newClassPathResource("file1.xml", this.getClass())
    kfs.write(ks.getResources().newClassPathResource("file2.xml", this.getClass())

    // Throws NPE
    KieBuilder kieBuilder = ks.newKieBuilder(kfs).buildAll();
}
```

### PR DESCRIPTION
**Pull request by samuel-beniamin**
Closes https://github.com/apache/incubator-kie-issues/issues/1474

### CODE REVIEW

**Comment by pibizza** [comment]
Added @gitgabrio, he is working on this stuff at the moment.

File: kie-dmn/kie-dmn-signavio/src/test/java/org/kie/dmn/signavio/SignavioTest.java [hunk 1]
```diff
@@ -291,31 +293,45 @@ void signavioIterateMultiinstanceWithComplexInputs() {
...
+        DMNRuntime dmnRuntime = createRuntime("MID with outside requirement.dmn", "survey MID SUM.dmn", "Signavio_Concat.dmn");
+
+        List<String> decisionNames = dmnRuntime
+                .getModels()
+                .stream()
+                .flatMap(dmnModel -> dmnModel.getDecisions().stream())
+                .map(DMNNode::getName)
+                .collect(Collectors.toList());
+
+        assertThat(decisionNames).containsOnly("sumUp", "iterating", "determineModifier", "concatNames");
```
Comment (pibizza) [inline kie-dmn/kie-dmn-signavio/src/test/java/org/kie/dmn/signavio/SignavioTest.java]: Maybe you can simply this and the previous statement using directly assertj assertions. Assertj has a powerful fluent interface to manipulate streams and perform stream operation. You can check extracting and flatExtracting, for example. 
```suggestion
        assertThat(dmnRuntime.getModels()).flatExtracting(dmnModel -> dmnModel.getDecisions()).extracting(DMNNode::getName).containsOnly("sumUp", "iterating", "determineModifier", "concatNames");
```

**Comment by gitgabrio** [comment]
Hi @samuel-beniamin 
Reading the original ticket, it seems there is some sort of issue in the implementation.
But IINW, this PR does not change the implementation at all (if just add a log), and rather modifies a test: could you clarify that ?

**Comment by samuel-beniamin** [comment]
> Hi @samuel-beniamin Reading the original ticket, it seems there is some sort of issue in the implementation. But IINW, this PR does not change the implementation at all (if just add a log), and rather modifies a test: could you clarify that ?

Hello @gitgabrio, It is in the implementation, I can still see my change returning or exiting early incase the model is not the same. 

Screenshot 2024-09-12 at 15 53 20

So basically, the compiler calls this callback regardless of the model it is trying to evaluate, but in this case the extra check is basically the fix, it seems trivial but it is blocking if we need to load multiple models.
```java
if (cModel != model) {
  return;
}
```

File: kie-dmn/kie-dmn-signavio/src/test/java/org/kie/dmn/signavio/SignavioTest.java (hunk 1 above)
Comment (samuel-beniamin) [reply kie-dmn/kie-dmn-signavio/src/test/java/org/kie/dmn/signavio/SignavioTest.java]: Looks good, applied ✅

**Comment by samuel-beniamin** [comment]
@gitgabrio So skipping the execution of the callback in the specific case of Signavio's extension and in the case it is a MID executing callback for a different model, is what you are working on right now?

**Comment by gitgabrio** [comment]
@samuel-beniamin first of all, thanks for the PR ❤️ 
And then, thanks for explanation - my bad, I was confused by the log.
I see your point, and this PR is fine for me.
But TBH, we did not touch this module for some time, so we would need to dig in it a bit; and as you wrote 

>  but maybe the proper fix should be in the DMNCompilerImpl.

this seems more a workaround for a symptom of a problem somewhere else.
So, if you may, I would kindly ask you to create another ticket (or duplicate the original one) so that we could focus on the root problem, eventually with your help : wdyt ?

**Review by gitgabrio — APPROVED** [review]

**Review by pibizza — APPROVED** [review]
Looks good to me.

**Review by gitgabrio — APPROVED** [review]

**Review by gitgabrio — APPROVED** [review]

**Review by yesamer — APPROVED** [review]

**Comment by samuel-beniamin** [comment]
@gitgabrio Thank you for approving the the PR.

> this seems more a workaround for a symptom of a problem somewhere else.

Yes, I agree too, I think a proper fix should be to validate for which model is the call back being executed, by the compiler itself. I will create another ticket for it.

**Comment by samuel-beniamin** [comment]
Here is a ticket to address a proper fix in the future https://github.com/apache/incubator-kie/issues/6974

**Comment by gitgabrio** [comment]
Great @samuel-beniamin , many thanks!!!

**Comment by samuel-beniamin** [comment]
Hey @gitgabrio, I am wondering how can I move this fix to be included in the next planned release? And what is the next planned release?

**Comment by gitgabrio** [comment]
Hi @samuel-beniamin 
this PR has been merged, and IINW it should already be in 10.0 release: am I missing something ?

**Comment by samuel-beniamin** [comment]
> Hi @samuel-beniamin this PR has been merged, and IINW it should already be in 10.0 release: am I missing something ?

I tried to check both on 
1. The released JAR file in maven for version 10.0.0 https://mvnrepository.com/artifact/org.kie/kie-dmn-signavio/10.0.0 and it didn't include the fix.
2. The branch 10.0.x it didn't include this fix or commit.

Assuming I am not wrong, what would be next release to include this commit? and what can I do or assist with from my side?
If I am wrong, could you please point me to the public release of the JAR file?

**Comment by gitgabrio** [comment]
Sorry @samuel-beniamin 
the 10 release has been slightly cumbersome due to the big changes we had in the meantime, so probably I'm confusing date.
@baldimir @yesamer, could you please help here ?

**Comment by yesamer** [comment]
@samuel-beniamin @gitgabrio 
IINW, we branched 10 version before @samuel-beniamin PR, as you may notice here --> https://github.com/apache/incubator-kie-drools/commits/10.0.x/
It seems that we created the 10 branch on Jul 16, 2024, and @samuel-beniamin's PR was not backported in that branch, as was merged on Sep 13, 2024.
I'm afraid you need to wait for the next release to have your changes released.
I guess you will not be happy about that @samuel-beniamin, I apologize for that, but as @gitgabrio said 10 release was a nightmare

**Comment by baldimir** [comment]
Hi, correct, what @yesamer wrote. @samuel-beniamin If you cannot wait for the next release, you could backport your fix locally to the 10.0.x branch and you could build a patched jar meanwhile. It should be just a matter of building with Maven (mvn clean install). If you need to use it with multiple people, you would need to upload it to your Maven repository or each person copy it to their m2 local repositories.

**Comment by samuel-beniamin** [comment]
Hello @baldimir, @yesamer and @gitgabrio 

First, thank you all for your answers. Unfortunately we have to wait in that case, as we are not the consumers of that library. Our customers are relying on it, and hence we can't just use the dependency as an m2 patch locally, but in fact wait for the next release. I am still wondering if the next release is planned for a date or even an estimated quarter? Any hint would be greatly appreciated.
Thank you all again 🙇

**Comment by yesamer** [comment]
Hi @samuel-beniamin, we started the discussion about the release of 10.1.0 in the previous week (that will contain your fix)
Despite a released date is still not defined, I would be surprised if 10.1.0 will land after 2025 Q1  . Please stay tuned in the next weeks :)

**Comment by gitgabrio** [comment]
👍

**Pull request merged**